ProcessLib/HeatConduction/HeatConductionFEM.h

```diff
@@ -0,0 +1,175 @@
+{
+    using ShapeMatricesType = ShapeMatrixPolicyType<ShapeFunction, GlobalDim>;
+    using ShapeMatrices = typename ShapeMatricesType::ShapeMatrices;
+
+    using LocalAssemblerTraits = ProcessLib::LocalAssemblerTraits<
+        ShapeMatricesType, ShapeFunction::NPOINTS, NUM_NODAL_DOF, GlobalDim>;
+
+    using NodalMatrixType = typename LocalAssemblerTraits::LocalMatrix;
+    using NodalVectorType = typename LocalAssemblerTraits::LocalVector;
+    using GlobalDimVectorType = typename ShapeMatricesType::GlobalDimVectorType;
+
+public:
+    /// The thermal_conductivity factor is directly integrated into the local
+    /// element matrix.
+    LocalAssemblerData(MeshLib::Element const& element,
+                       std::size_t const local_matrix_size,
+                       unsigned const integration_order,
+                       HeatConductionProcessData const& process_data)
+        : _element(element),
+          _process_data(process_data),
+          _localK(local_matrix_size,
+                  local_matrix_size),  // TODO narrowing conversion
+          _localM(local_matrix_size, local_matrix_size),
+          _localRhs(local_matrix_size),
+          _integration_method(integration_order),
+          _shape_matrices(initShapeMatrices<ShapeFunction, ShapeMatricesType,
+                                            IntegrationMethod, GlobalDim>(
+              element, _integration_method))
+    {
+        // This assertion is valid only if all nodal d.o.f. use the same shape
+        // matrices.
+        assert(local_matrix_size == ShapeFunction::NPOINTS * NUM_NODAL_DOF);
+    }
+
+    void assembleConcrete(
+        double const t, std::vector<double> const& local_x,
+        NumLib::LocalToGlobalIndexMap::RowColumnIndices const& indices,
+        GlobalMatrix& M, GlobalMatrix& K, GlobalVector& b) override
+    {
+        _localK.setZero();
+        _localM.setZero();
+        _localRhs.setZero();
+
+        unsigned const n_integration_points =
+            _integration_method.getNumberOfPoints();
+
+        SpatialPosition pos;
+        pos.setElementID(_element.getID());
+
+        for (unsigned ip = 0; ip < n_integration_points; ip++)
+        {
+            pos.setIntegrationPoint(ip);
+            auto const& sm = _shape_matrices[ip];
+            auto const& wp = _integration_method.getWeightedPoint(ip);
+            auto const k = _process_data.thermal_conductivity(t, pos)[0];
+            auto const heat_capacity = _process_data.heat_capacity(t, pos)[0];
+            auto const density = _process_data.density(t, pos)[0];
+
+            _localK.noalias() +=
+                sm.dNdx.transpose() * k * sm.dNdx * sm.detJ * wp.getWeight();
+            _localM.noalias() += sm.N.transpose() * density * heat_capacity *
+                                 sm.N * sm.detJ * wp.getWeight();
+            // heat flux only computed for output.
+            const NodalVectorType heat_flux =
```

Comment:
@endJunction I have tested the error threshold. 1e-4 is required for neumann boundary  and 1e-5 is fine with dirichlet boundary ! I have updated the threshold.